### ISSUE TITLE
jax.autograd very slow, and jax.numpy slower than standard numpy

### ISSUE THREAD

**Issue by apal90** [google/jax#1318]
I am not sure if this is the correct forum for this issue, so please correct me if I am wrong. I just started using jax in order to implement its GPU-backed autograd capability. However, for my function(s), the gradient is either extremely slow to calculate, or worse, makes the kernel crash (always the case with `jit` on). Moreover, even the original scalar function (and functions used within) is many times slower when used with `jax.numpy` rather than standard numpy. Using `jit` significantly quickens the scalar function, but that does not affect the performance of the gradient.

Can someone please point out to me any significant bottlenecks in my code, perhaps things that are unsuited to jax? The code is given below. The scalar function is `bendE__triangles()`, which calls upon `bendEdensity__triangles()` and `getCurvatures()`.  This last contains all the complicated operations. The inputs are all standard numpy ndarrays.

```
import jax.numpy as np
from jax import grad, vmap, jit
from jax.config import config
config.update("jax_enable_x64", True)
```

```
@jit 
def getCurvatures(xyz, triSimplicesInterior, triNL, W):
    ## initialise arrays
    xyzTri = np.zeros((6,3), dtype=np.float64) 
    zProjected = np.zeros(6, dtype=np.float64) 
    centroid = np.zeros(3, dtype=np.float64) 
    aVector = np.zeros(6, dtype=np.float64) 
    curvVector = []
    
    for t, triangle in enumerate(triSimplicesInterior):

        p1, p2, p3 = triangle[0], triangle[1], triangle[2]
        
        x1, y1, z1 = xyz[p1][0], xyz[p1][1], xyz[p1][2]
        x2, y2, z2 = xyz[p2][0], xyz[p2][1], xyz[p2][2]
        x3, y3, z3 = xyz[p3][0], xyz[p3][1], xyz[p3][2]
        
        N1 = np.cross(xyz[p2]-xyz[p1] , xyz[p3]-xyz[p2])
        n1 = N1 / np.linalg.norm(N1)
  
        xyzTri = np.vstack((xyz[triangle], xyz[triNL[t]])) 
        
        centroid = np.array([(x1+x2+x3)/3, (y1+y2+y3)/3, (z1+z2+z3)/3])
        zProjected = np.dot(xyzTri-centroid, n1)

        aVector = np.dot(W[t], zProjected)

        c11 = 2*aVector[3]
        c12 = aVector[4]
        c22 = 2*aVector[5]
        curvVector.append(np.array([c11,c12,c22]))
    
    curvVector = np.array(curvVector)
    return curvVector

@jit
def bendEdensity__triangles(xyz0, triSimplicesInterior, triNL, W, triAreasInterior):
    
    xyz = np.reshape(xyz0, (xyz0.size//3, 3))

    curvMatrix = getCurvatures(xyz, triSimplicesInterior, triNL, W)

    c11, c22 = 0., 0.
    energy = 0.
    
    bendEdensity = []
    for t in range(curvMatrix.shape[0]):
        c11, c22 = curvMatrix[t][0], curvMatrix[t][2]
        bendEdensity.append(0.5*(c11+c22)**2) # = 1/2(TraceC)**2 
    
    bendEdensity = np.array(bendEdensity)
    return bendEdensity

 @jit
 def bendE__triangles(xyz0, triSimplicesInterior, triNL, W, triAreasInterior):
    
    xyz = np.reshape(xyz0, (xyz0.size//3, 3))

    bendEdensity = bendEdensity__triangles(xyz, triSimplicesInterior, triNL, W, triAreasInterior)
    energy = np.sum(bendEdensity * triAreasInterior) # = Sum( 1/2(TraceC)**2 * area) 

    return energy


## scalar energy function
def bendE(xyz):
        return bendE__triangles(xyz, triSimplicesInterior, triNLinterior, W0, triAreasInterior)

## gradient using jax.grad
der_energy = grad(bendE)

print(bendE(xyz0)) # speeded up by using jit
print(der_energy (xyz0)) # either takes long time, or kills the kernel.
```

**Comment by mattjj** [comment]
This is the perfect forum for this kind of issue! So thanks for raising it here :)

We like things to be fast, and if your example is self-contained that will let us dig in and make JAX better. We'll get back to you when we take a look!

**Comment by hawkinsp** [comment]
Although I note the example isn't self-contained. Can you make it self-contained, i.e., something that we can run without guessing what inputs to provide, etc.? You can probably use fake data rather than real data to illustrate this kind of performance problem.

I also don't see any gradients in the code? Can you show code that exhibits the gradient performance problem?

**Comment by sschoenholz** [comment]
Glancing at this code, it seems like there are a lot of per-triangle / per-vertex operations that are likely the origin of the slow behavior (and, to me, it makes sense that JAX would be slow here since GPUs are advantageous only when you can process a lot of content in parallel). You would probably find it much faster to vectorize the calculations. While you could do this manually, I've found it very nice to write code to act on e.g. one triangle or one simplex and then vectorize use `vmap`. Do you think that approach could work here?

**Comment by apal90** [comment]
That sounds appealing, but a quick look at the code doesn't reveal any places where I could vectorise easily. For example, the calculations on the simplices involve the neighbouring simplices. But then again, I am not a very experienced coder; perhaps someone else can notice easily vectorisable spots that escape me. I will keep looking though.

**Comment by jekbradbury** [comment]
`[partialFunc(x, y) for x,y in zip(triSimplicesInterior, triNLinterior)]` corresponds to `vmap(partialFunc, in_axes=(0, 0))(triSimplicesInterior, triNLinterior)`.

**Comment by sschoenholz** [comment]
You can even drop the in_axes, `vmap(partialFunc)(triSimplicesInterior, triNLinterior)`.

**Comment by apal90** [comment]
@sschoenholz @jekbradbury  I believe I almost have it. I realised that I in fact need to iterate over three arrays, with different shapes but same first dimension:
`[partialFunc(x, y, z) for x,y,z in zip(triSimplicesInterior, triNLinterior, W)]`

where `triSimplicesInterior` has shape `N,3`, `triNLinterior` has shape `N,3`, and `W` has shape `N,6,6`.

While the above list comprehension works, using `vmap` naively does not work. 

`vmap(partialFunc)(triSimplicesInterior, triNLinterior, W)` gives the error:

```
Exception: Tracer can't be used with raw numpy functions. You might have
  import numpy as np
instead of
  import jax.numpy as np
```

What do I do here?

**Comment by jekbradbury** [comment]
You need to make sure `partialFunc` is made entirely of `jax.numpy` functions and doesn't have any `onp` functions in it.

**Comment by apal90** [comment]
> You need to make sure `partialFunc` is made entirely of `jax.numpy` functions and doesn't have any `onp` functions in it.

So, it's not a problem due to differing array shapes/ndim? Do the `onp.ndarrays` also need to be redefined as `jax.numpy` arrays?

**Comment by sschoenholz** [comment]
+1 to James. Perhaps you can set up an example in a colab notebook? I'd be happy to take a glance at it.

**Comment by apal90** [comment]
@sschoenholz  @jekbradbury  Thanks for that idea. Here is the link: https://colab.research.google.com/drive/1VCPiRTLfxOflooDtTlqHI4VzYQkXzaUn

All the functions now run perfectly. I just needed to restart the kernel for the errors to go away. However, I am now getting a more complex error: the scalar values are as expected, but the gradient is now a zero matrix!  

So, in the colab notebook, I have put in three scenarios.

1. The original code (pre-vectorisation), with gradient calculated using `autograd.grad`. This gives values for the scalar that I trust, and also a gradient that I feel that I can trust. At least, it's non-zero.
2. Vectorised code, using `autograd.numpy` and `autograd.grad`. I get back the same scalar values, but a zero gradient.
3. Vectorised code, using `jax`. Again, I get back the same scalar values, but a zero gradient.

So, something happened during the vectorisation procedure that prevents both `autograd` and `jax` from calculating the true gradient. Can you figure out what?

**Comment by apal90** [comment]
@sschoenholz  @jekbradbury Could you please take a look at the code in the notebook when you can? I would really like to solve this zero gradient issue.

**Comment by mattjj** [comment]
I just tried running the repro notebook (yes, just over 3 years later!). There was an issue with the code; the value of `index_update` wasn't being used. That explains the zero gradients!

I edited the code by changing this line:
```python
    index_update(curvVector, 1, aVector[4]) # c12 = aVector[4]
```

to this:
```python
    curvVector = curvVector.at[1].set(aVector[4])
```

(The Autograd code, i.e. the code using https://github.com/hips/autograd, had a similar issue.)

Then on CPU I got these timings:
<img width="831" alt="image" src="https://user-images.githubusercontent.com/1458824/191592495-ea7b0b95-e36c-4f2e-a3f6-dee361d82a44.png">

I think the primary problem here was how to write indexed assignment using JAX, which has (since 2019) improved substantially. I'm sure there could've been more iteration in 2019 (optimizing performance etc), but given that we've figured out the primary problem, I'm going to close this issue.
